File: codes/supporting_codes/stim_waveform.m
function phi_x = stim_waveform(step, params)
% stim waveform generator.
% Default: legacy burst/pulse iTBS/cTBS logic (your existing implementation).
% Optional: continuous sinusoidal pulse-train when params.stim.mode = 'continuous'.

    % Convert step index to time 
    t = (step-1) * params.deltat;

    % -------------------------------
    % NEW: continuous mode (early return)
    % -------------------------------
    if isfield(params,'stim') && isfield(params.stim,'mode') ...
            && strcmpi(params.stim.mode,'continuous')

        % Use dedicated helper. This must return a SCALAR.
        phi_x = continuous_stim_waveform(step, params);

        % IMPORTANT: return now so legacy burst logic cannot overwrite it
        return;
    end

    % -------------------------------
    % Legacy burst/pulse mode (UNCHANGED)
    % -------------------------------

    if strcmp(params.stim_mode, 'intermittent')

        dt = params.deltat;
        t  = (step - 1) * dt;

        if t < params.onset
            phi_x = 0;
            return;
        end

        % Integer sample index since onset
        k = round((t - params.onset) / dt);

        duration_steps = round(params.Duration / dt);

        % Use [onset, onset + Duration), excluding endpoint
        if k < 0 || k >= duration_steps
            phi_x = 0;
            return;
        end

        pulse_width_steps  = round(params.width / dt);
        pulse_period_steps = round((1 / params.freq) / dt);
        burst_period_steps = round((1 / params.oscillation_freq) / dt);

        if params.off == 0
            k_in_on = k;
        else
            on_steps    = round(params.on / dt);
            cycle_steps = round((params.on + params.off) / dt);

            k_cycle = mod(k, cycle_steps);

            if k_cycle >= on_steps
                phi_x = 0;
                return;
            end

            k_in_on = k_cycle;
        end

        k_in_burst = mod(k_in_on, burst_period_steps);

        pulse_idx  = floor(k_in_burst / pulse_period_steps);
        k_in_pulse = k_in_burst - pulse_idx * pulse_period_steps;

        if pulse_idx < params.bursts && k_in_pulse < pulse_width_steps
            phi_x = params.amp;
        else
            phi_x = 0;
        end

        return;
    end

    error('Unknown params.stim_mode: %s', params.stim_mode);
end

File: codes/py_codes/stim_waveform.py
import numpy as np
from continuous_stim_waveform import continuous_stim_waveform


def stim_waveform(step, params):
    """
    Python translation of stim_waveform.m.

    If params.stim.mode == 'continuous', this calls continuous_stim_waveform()
    and returns immediately.

    Otherwise, it uses integer-index intermittent/burst pulse logic to avoid
    floating-point boundary artifacts. This is the corrected replacement for
    the legacy MATLAB floating-time/mod pulse logic.
    """
    stim = getattr(params, 'stim', None)
    if stim is not None and getattr(stim, 'mode', '').lower() == 'continuous':
        return continuous_stim_waveform(step, params)

    return intermittent_stim_waveform_integer(step, params)


def intermittent_stim_waveform_integer(step, params):
    """
    Corrected intermittent/burst waveform using integer sample indices.

    This branch keeps the old TMS/stim convention:
        t = (step - 1) * params.deltat

    Stimulation is active on [onset, onset + Duration), excluding the endpoint.
    Therefore no one-sample trailing pulse is generated at onset + Duration.
    """
    dt = params.deltat
    t = (step - 1) * dt

    if t < params.onset:
        return 0.0

    k = int(round((t - params.onset) / dt))
    duration_steps = int(round(params.Duration / dt))

    if k < 0 or k >= duration_steps:
        return 0.0

    pulse_width_steps = int(round(params.width / dt))
    pulse_period_steps = int(round((1.0 / params.freq) / dt))
    burst_period_steps = int(round((1.0 / params.oscillation_freq) / dt))

    if getattr(params, 'off', 0.0) == 0:
        k_in_on = k
    else:
        on_steps = int(round(params.on / dt))
        cycle_steps = int(round((params.on + params.off) / dt))
        k_cycle = k % cycle_steps
        if k_cycle >= on_steps:
            return 0.0
        k_in_on = k_cycle

    k_in_burst = k_in_on % burst_period_steps
    pulse_idx = k_in_burst // pulse_period_steps
    k_in_pulse = k_in_burst - pulse_idx * pulse_period_steps

    if pulse_idx < params.bursts and k_in_pulse < pulse_width_steps:
        return float(params.amp)

    return 0.0
